client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java

```diff
@@ -1462,25 +1458,23 @@ public Message request(final Message msg, final MessageQueueSelector selector, f
         final String correlationId = msg.getProperty(MessageConst.PROPERTY_CORRELATION_ID);
 
         try {
-            final RequestResponseFuture requestResponseFuture = new RequestResponseFuture(correlationId, timeout, null);
-            RequestFutureHolder.getInstance().getRequestFutureTable().put(correlationId, requestResponseFuture);
+            final RequestResponseFuture responseFuture = new RequestResponseFuture(correlationId, timeout, null);
+            RequestFutureHolder.getInstance().getRequestFutureTable().put(correlationId, responseFuture);
 
             long cost = System.currentTimeMillis() - beginTimestamp;
             this.sendSelectImpl(msg, selector, arg, CommunicationMode.ASYNC, new SendCallback() {
                 @Override
                 public void onSuccess(SendResult sendResult) {
-                    requestResponseFuture.setSendRequestOk(true);
+                    requestSuccess(correlationId, new Message(msg.getTopic(), msg.getBody()));
                 }
 
                 @Override
                 public void onException(Throwable e) {
-                    requestResponseFuture.setSendRequestOk(false);
-                    requestResponseFuture.putResponseMessage(null);
-                    requestResponseFuture.setCause(e);
+                    requestFail(correlationId, e);
                 }
             }, timeout - cost);
 
-            return waitResponse(msg, timeout, requestResponseFuture, cost);
+            return waitResponse(msg, timeout - cost, responseFuture, cost);
```

Comment:
`timeout` or `timeout - cost` ?

Reply:
> time - cost More accurate, cost is pre handler time

+1

Reply:
enen, but the scope of current PR modification is the request method